utils/contextutils/context.go

```diff
@@ -0,0 +1,65 @@
+// Package contextutils provides utility for adding and retrieving metadata to/from a context.
+package contextutils
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+type contextKey string
+
+const (
+	// MetadataKey is the key used to access metadata from a context with metadata.
+	MetadataKey = contextKey("viam-metadata")
+
+	// TimeRequestedMetadataKey is optional metadata in the gRPC response header that correlates
+	// to the time right before the point cloud was captured.
+	TimeRequestedMetadataKey = "viam-time-requested"
+
+	// TimeReceivedMetadataKey is optional metadata in the gRPC response header that correlates
+	// to the time right after the point cloud was captured.
+	TimeReceivedMetadataKey = "viam-time-received"
+)
+
+// ContextWithMetadata attaches a metadata map to the context.
+func ContextWithMetadata(ctx context.Context) (context.Context, map[string][]string) {
+	// If the context already has metadata, return that and leave the context untouched.
+	existingMD := ctx.Value(MetadataKey)
+	if mdMap, ok := existingMD.(map[string][]string); ok {
+		return ctx, mdMap
+	}
+
+	// Otherwise, add a metadata map to the context.
+	md := make(map[string][]string)
+	ctx = context.WithValue(ctx, MetadataKey, md)
+	return ctx, md
+}
+
+// ContextWithMetadataUnaryClientInterceptor attempts to read metadata from the gRPC header and
+// injects the metadata into the context if the caller has passed in a context with metadata.
+func ContextWithMetadataUnaryClientInterceptor(
+	ctx context.Context,
+	method string,
+	req, reply interface{},
+	cc *grpc.ClientConn,
+	invoker grpc.UnaryInvoker,
+	opts ...grpc.CallOption,
+) error {
+	var header metadata.MD
+	opts = append(opts, grpc.Header(&header))
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err != nil {
+		return err
+	}
+
+	md := ctx.Value(MetadataKey)
+	if mdMap, ok := md.(map[string][]string); ok {
```

Comment:
Should an error be returned if `ContextWithMetadataUnaryClientInterceptor` and no `md` in the `map[string][]string` format is available?